### PR TITLE
Adds unit test to check Exit Code

### PR DESCRIPTION
**Pull request by beholders-eye**
ExecShell isn't returning an error nor an exit code != 0
The CLI isn't working as expected because of it

Also updated go-dockerclient to the latest upstream version from fsouza

### CODE REVIEW

**Comment by beholders-eye** [comment]
> Code looks good. There were a ton of changes between John's fork and the current go-client. Is this when you want to update the library?

I thought it would fix the issue. This new unit test always fail, while it shouldn't, right?
My question here is: I'm expecting something wrong?

**Comment by beholders-eye** [comment]
Just to add more context, as @jamesnaftel said: upstream go-dockerclient, from fsouza is 160 commits ahead of John's fork. I updated it and it built just alright. But the error doesn't come out still.

**Comment by zombiezen** [comment]
> Code looks good. There were a ton of changes between John's fork and the current go-client. Is this when you want to update the library?

I'm not sure the differences are too large: https://github.com/fsouza/go-dockerclient/compare/main...johnewart:master

As a minor sidebar: this is going to be a breaking change for users of the Narwhal package, since we're changing the type we pass into the functions. It's an easy-to-fix breakage, since we just change an import, but it means we should bump minor version, not just the patch version.

File: containers_test.go [hunk 1]
```diff
@@ -30,6 +30,45 @@ func TestMain(m *testing.M) {
...
+		t.Fatal(err)
+	}
+
+	cmdStat := "stat /somewhere/strange/that/do/not/exist"
+	err = ExecShell(ctx, client, container.ID, cmdStat, &ExecShellOptions{
+		Interactive: true,
+	})
+	if err != nil {
+		if code, ok := IsExitError(err); !ok {
+			t.Errorf("Exit code: want 1, got %d", code)
```
Comment (zombiezen) [inline containers_test.go:51]: If the error doesn't match, the full error message is probably more helpful here:

```suggestion
		if code, ok := IsExitError(err); !ok || code != 1 {
			t.Errorf("error = %v; want exit code 1", err)
```

Comment (beholders-eye) [reply containers_test.go:51]: It's possibly that I've hit moby/moby/issues/34488, still checking

Comment (beholders-eye) [reply containers_test.go:51]: But I'm running 19.03.8-ce runc/moby, so I'm not entirely sure

**Comment by beholders-eye** [comment]
PTAL

Don't know exactly why, but if ExecShell doesn't get an output, even a simple buffer, it won't return the exit code nor an error, it ignores the whole execution, in both situations, where we try to `mkdir <something/something>` (without `-p`) and `stat <somewhere/crazy>`

File: containers_test.go [hunk 2]
```diff
@@ -30,6 +30,68 @@ func TestMain(m *testing.M) {
...
+	err = ExecShell(ctx, client, containerID, cmdStat, &ExecShellOptions{
+		CombinedOutput: os.Stdout,
+	})
+	if err != nil {
+		if code, ok := IsExitError(err); !ok || code != 1 {
+			t.Errorf("error = %v; want exit code 1", err)
+		}
+	} else {
+		t.Error("Expecting an error, but none was returned")
+	}
```
Comment (zombiezen) [inline containers_test.go:78]: The `err != nil` check is not necessary: `IsExitError` is sufficient to check.

```suggestion
	if code, ok := IsExitError(err); !ok || code != 1 {
		t.Errorf("error = %v; want exit code 1", err)
	}
```

File: containers_test.go [hunk 3]
```diff
@@ -30,6 +30,68 @@ func TestMain(m *testing.M) {
...
+	err = ExecShell(ctx, client, containerID, cmdMkdir, &ExecShellOptions{
+		CombinedOutput: os.Stdout,
+	})
+	if err != nil {
+		if code, ok := IsExitError(err); !ok || code != 1 {
+			t.Errorf("error = %v; want exit code 1", err)
+		}
+	} else {
+		t.Error("Expecting an error, but none was returned")
+	}
```
Comment (zombiezen) [inline containers_test.go:91]: Same here:

```suggestion
	if code, ok := IsExitError(err); !ok || code != 1 {
		t.Errorf("error = %v; want exit code 1", err)
	}
```

**Comment by zombiezen** [comment]
> Don't know exactly why, but if ExecShell doesn't get an output, even a simple buffer, it won't return the exit code nor an error, it ignores the whole execution, in both situations, where we try to `mkdir <something/something>` (without `-p`) and `stat <somewhere/crazy>`

Interesting! Since this is fairly surprising behavior, perhaps we should make the `ExecShell` function always set an output, even if the argument is `nil`. Something like [`ioutil.Discard`](https://pkg.go.dev/io/ioutil?tab=doc#Discard) should work.

File: containers_test.go (hunk 2 above)
Comment (beholders-eye) [reply containers_test.go:78]: If err is `nil`, `IsExitError` won't be able to treat anything, so no, it should be kept as is. When the test was failing (without a `CombinedOutput`), it was showing "Expecting an error, but none was returned".

Comment (zombiezen) [reply containers_test.go:78]: I don't understand the objection. If I run:

```go
package main

import (
	"fmt"
	"github.com/yourbase/narwhal"
)

func main() {
	fmt.Println(narwhal.IsExitError(nil))
}
```

The output is:

```
0 false
```

So `!ok || code != 1` would be `true`, and thus trigger a test failure.

Comment (beholders-eye) [reply containers_test.go:78]: It gave me an error, locally. Maybe I did something wrong. I'll check back

**Pull request merged**